sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala

```diff
@@ -48,6 +49,33 @@ case class Mode(
 
   override def inputTypes: Seq[AbstractDataType] = Seq(AnyDataType)
 
+  override def checkInputDataTypes(): TypeCheckResult = {
+    checkDataType(child.dataType)
+  }
+
+  private def checkDataType(dataType: DataType, level1: Boolean = true): TypeCheckResult = {
+    dataType match {
+      case ArrayType(elementType, _) =>
+        checkDataType(elementType, level1 = false)
+      case StructType(fields) =>
+        combineTypeCheckResults(fields.map { field =>
+          checkDataType(field.dataType, level1 = false)
+        })
+      case dt: StringType if !level1 &&
+        !CollationFactory.fetchCollation(dt.collationId).supportsBinaryEquality
+      => TypeCheckResult.TypeCheckFailure(
+        s"Input to function $prettyName was a complex type" +
+          s" with strings collated on non-binary collations," +
+          s" which is not yet supported.")
+      case _ => TypeCheckResult.TypeCheckSuccess
+    }
+  }
```

Comment:
instead of this recursive check, consider using already existing recursive constructs (such as `existsRecursively` in `DataType`)